daanmlab/ft_transcendence
# [TID-66]Feature/game invitation

Created game invitation system. 
- Added GameInvitation model
- API endpoints
- WebSocket communication for sending and receiving invitations
- Frontend flow

To test:

API: 
- Make friends
- At /onevsone make some invitations
- Log in with the invited user. Receive and accept invitations.
- Check Invitation and Game creation on the models and database

WebSockets:
- Receive invitations while logged in, accept invitation
- Receive acceptance notifications while logged in
- In both cases, user should be redirected to game page
- Check Invitation and Game creation on the models and database

------------------------------------------------------

### Backend Changes:

* [`backend/app/consumers.py`](diffhunk://#diff-af411d764b50f9e7a814158f9b1f9ffee527a1b043edca4d56db2e04fb6d9215R1-R35): Added `GameInvitationConsumer` for WebSocket communication to handle game invitations.
* [`backend/app/models.py`](diffhunk://#diff-96c5a7997e3d147589c1d6cac9cf36730df0fe459ae0170a2fc90ededd63fed2R29-R44): Introduced `GameInvitation` model to manage game invitations with status tracking.
* [`backend/app/routing.py`](diffhunk://#diff-e696e5f0d2eb9d3e67e721e77c135e69c139fee74169c43449147ac1f0fd2593R4-R10): Updated WebSocket routing to include the new `GameInvitationConsumer`.
* [`backend/app/urls.py`](diffhunk://#diff-5df129349fd1329ba4d60780a35b3da68bfd38819e97d12d754563a9b1bf44bcR27-R42): Added new URL patterns for creating, accepting, and listing game invitations. [[1]](diffhunk://#diff-5df129349fd1329ba4d60780a35b3da68bfd38819e97d12d754563a9b1bf44bcR27-R42) [[2]](diffhunk://#diff-5df129349fd1329ba4d60780a35b3da68bfd38819e97d12d754563a9b1bf44bcL38-R55)
* [`backend/app/views/game_views.py`](diffhunk://#diff-7998cedd5e041c99502a153e80d98f4e2f73270838d23a48a5c4df7498180521R1-R98): Implemented views for creating and accepting game invitations, and listing sent/received invitations.

### Frontend Changes:

* [`frontend/src/js/Api.js`](diffhunk://#diff-baefc9bb54df6008c7fbdede5e079177c7017e8ae4cd6b03165983144294794eR87-R94): Added methods to handle game invitations, including sending, accepting, and listing invitations. [[1]](diffhunk://#diff-baefc9bb54df6008c7fbdede5e079177c7017e8ae4cd6b03165983144294794eR87-R94) [[2]](diffhunk://#diff-baefc9bb54df6008c7fbdede5e079177c7017e8ae4cd6b03165983144294794eL119-L143) [[3]](diffhunk://#diff-baefc9bb54df6008c7fbdede5e079177c7017e8ae4cd6b03165983144294794eL154) [[4]](diffhunk://#diff-baefc9bb54df6008c7fbdede5e079177c7017e8ae4cd6b03165983144294794eL163-R205)
* [`frontend/src/js/Auth.js`](diffhunk://#diff-6b2a96a0c5daaaf21054fea24143a9c9e0261645680d4ffe5d7b9d54628d248cR110): Ensured WebSocket connection is closed upon logout. [[1]](diffhunk://#diff-6b2a96a0c5daaaf21054fea24143a9c9e0261645680d4ffe5d7b9d54628d248cR110) [[2]](diffhunk://#diff-6b2a96a0c5daaaf21054fea24143a9c9e0261645680d4ffe5d7b9d54628d248cR215-R219)
* [`frontend/src/js/index.js`](diffhunk://#diff-4806696a2f600c50cabcf2249c4d60723d6db86f8919464f15d6ef418c4e04daR1): Integrated new `GamePage` and WebSocket connection handling for game invitations. [[1]](diffhunk://#diff-4806696a2f600c50cabcf2249c4d60723d6db86f8919464f15d6ef418c4e04daR1) [[2]](diffhunk://#diff-4806696a2f600c50cabcf2249c4d60723d6db86f8919464f15d6ef418c4e04daL13-R15) [[3]](diffhunk://#diff-4806696a2f600c50cabcf2249c4d60723d6db86f8919464f15d6ef418c4e04daL39-R46)
